Teach process-css-properties.py to generate setters for hasExplicitlySet* flags

weinig opened the pull request:
#### 80ecd47209a9238675c1f9f61abf60e51101207a
<pre>
Teach process-css-properties.py to generate setters for hasExplicitlySet* flags
<a href="https://bugs.webkit.org/show_bug.cgi?id=212506">https://bugs.webkit.org/show_bug.cgi?id=212506</a>

Reviewed by Darin Adler.

Adds support to process-css-properties.py to generate code to set the
`hasExplicitlySet*` flags in `Style::BuilderGenerated`.

There are currently three different policies defined for when to set
the `hasExplicitlySet*` flags, configurable per-property via a new
entry &quot;render-style-has-explicitly-set-policy&quot; in CSSProperty.json.

The three policies are:
  - &quot;all-author-origin&quot;
      Invokes `builderState.style().setHasExplicitlySet*(builderState.isAuthorOrigin())`
      for `applyValue*`, `applyInitial*` and `applyInherit*`.

      Used by the `color` and `padding-*` properties.

  - &quot;all-border-radius&quot;
      Invokes `builderState.style().setHasExplicitlySet*(true)` for `applyValue*`.
      Invokes `builderState.style().setHasExplicitlySet*(false)` for `applyInitial*`.
      Invokes `builderState.style().setHasExplicitlySet*(builderState.parentStyle().hasExplicitlySet*())` for `applyInherit*`.

      Used by the `border-*-radius` properties.

  - &quot;value-only&quot;
      Invokes `builderState.style().setHasExplicitlySet*(true)` for `applyValue*`.
      Does not change the flag at all for `applyInitial*` or `applyInherit*`.

      Used by the `direction`, `writing-mode`, `color-scheme`, `stroke-color`
      and `stroke-width` properties.

* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/scripts/process-css-properties.py:
* Source/WebCore/css/scripts/test/TestCSSProperties.json:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/CSSPropertyInitialValuesGeneratedInlines.h:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/CSSPropertyNames.gperf:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/CSSPropertyNames.h:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/CSSPropertyParsing.cpp:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/CSSStyleProperties+PropertyNames.idl:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/RenderStyleProperties+GettersInlines.h:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/RenderStyleProperties+SettersInlines.h:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/RenderStyleProperties.h:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/StyleBuilderGenerated.cpp:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/StyleChangedAnimatablePropertiesGenerated.cpp:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/StyleComputedStyleProperties+GettersInlines.h:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/StyleComputedStyleProperties+InitialInlines.h:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/StyleComputedStyleProperties+SettersInlines.h:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/StyleComputedStyleProperties.h:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/StyleExtractorGenerated.cpp:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/StyleInterpolationWrapperMap.cpp:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/StylePropertyShorthandFunctions.cpp:
* Source/WebCore/style/StyleBuilderCustom.h:
(WebCore::Style::BuilderCustom::applyValueDirection): Deleted.
(WebCore::Style::BuilderCustom::applyValueWritingMode): Deleted.
(WebCore::Style::BuilderCustom::applyValueColorScheme): Deleted.
(WebCore::Style::BuilderCustom::applyInitialBorderBottomLeftRadius): Deleted.
(WebCore::Style::BuilderCustom::applyInheritBorderBottomLeftRadius): Deleted.
(WebCore::Style::BuilderCustom::applyValueBorderBottomLeftRadius): Deleted.
(WebCore::Style::BuilderCustom::applyInitialBorderBottomRightRadius): Deleted.
(WebCore::Style::BuilderCustom::applyInheritBorderBottomRightRadius): Deleted.
(WebCore::Style::BuilderCustom::applyValueBorderBottomRightRadius): Deleted.
(WebCore::Style::BuilderCustom::applyInitialBorderTopLeftRadius): Deleted.
(WebCore::Style::BuilderCustom::applyInheritBorderTopLeftRadius): Deleted.
(WebCore::Style::BuilderCustom::applyValueBorderTopLeftRadius): Deleted.
(WebCore::Style::BuilderCustom::applyInitialBorderTopRightRadius): Deleted.
(WebCore::Style::BuilderCustom::applyInheritBorderTopRightRadius): Deleted.
(WebCore::Style::BuilderCustom::applyValueBorderTopRightRadius): Deleted.
(WebCore::Style::BuilderCustom::applyValueStrokeWidth): Deleted.
(WebCore::Style::BuilderCustom::applyValueStrokeColor): Deleted.
(WebCore::Style::BuilderCustom::applyInitialColor): Deleted.
(WebCore::Style::BuilderCustom::applyInheritColor): Deleted.
(WebCore::Style::BuilderCustom::applyInitialPaddingBottom): Deleted.
(WebCore::Style::BuilderCustom::applyInheritPaddingBottom): Deleted.
(WebCore::Style::BuilderCustom::applyValuePaddingBottom): Deleted.
(WebCore::Style::BuilderCustom::applyInitialPaddingLeft): Deleted.
(WebCore::Style::BuilderCustom::applyInheritPaddingLeft): Deleted.
(WebCore::Style::BuilderCustom::applyValuePaddingLeft): Deleted.
(WebCore::Style::BuilderCustom::applyInitialPaddingRight): Deleted.
(WebCore::Style::BuilderCustom::applyInheritPaddingRight): Deleted.
(WebCore::Style::BuilderCustom::applyValuePaddingRight): Deleted.
(WebCore::Style::BuilderCustom::applyInitialPaddingTop): Deleted.
(WebCore::Style::BuilderCustom::applyInheritPaddingTop): Deleted.
(WebCore::Style::BuilderCustom::applyValuePaddingTop): Deleted.
* Tools/Scripts/webkitpy/style/checkers/jsonchecker.py:

Canonical link: <a href="https://commits.webkit.org/305011@main">https://commits.webkit.org/305011@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2316fd03e8ceefd2e47844054b49349091a3c3da

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/137179 "13 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/9539 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/48466 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/144925 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/90152 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/139051 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/10242 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/9666 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/104906 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/76511 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/140124 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/7544 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/122931 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/85745 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/136527 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/7181 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/4891 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/5515 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/116534 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/41096 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/147684 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/9222 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/41658 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/113264 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/9240 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/7756 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/113595 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28849 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/7101 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/119188 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/63638 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/9270 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/37242 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/8996 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/72836 "Built successfully") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/9211 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/9063 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler reviewed (approved):
One way to confirm these policies are valuable for interoperability is to find a way to add clarifying tests or apecific wording to the web platform tests or CSS specifications or both.

Importance for compatibility may be a little harder to determine. But tests could even help indirectly with that.

weinig committed:
Teach process-css-properties.py to generate setters for hasExplicitlySet* flags
https://bugs.webkit.org/show_bug.cgi?id=212506

Reviewed by Darin Adler.

Adds support to process-css-properties.py to generate code to set the
`hasExplicitlySet*` flags in `Style::BuilderGenerated`.

There are currently three different policies defined for when to set
the `hasExplicitlySet*` flags, configurable per-property via a new
entry "render-style-has-explicitly-set-policy" in CSSProperty.json.

The three policies are:
  - "all-author-origin"
      Invokes `builderState.style().setHasExplicitlySet*(builderState.isAuthorOrigin())`
      for `applyValue*`, `applyInitial*` and `applyInherit*`.

      Used by the `color` and `padding-*` properties.

  - "all-border-radius"
      Invokes `builderState.style().setHasExplicitlySet*(true)` for `applyValue*`.
      Invokes `builderState.style().setHasExplicitlySet*(false)` for `applyInitial*`.
      Invokes `builderState.style().setHasExplicitlySet*(builderState.parentStyle().hasExplicitlySet*())` for `applyInherit*`.

      Used by the `border-*-radius` properties.

  - "value-only"
      Invokes `builderState.style().setHasExplicitlySet*(true)` for `applyValue*`.
      Does not change the flag at all for `applyInitial*` or `applyInherit*`.

      Used by the `direction`, `writing-mode`, `color-scheme`, `stroke-color`
      and `stroke-width` properties.

* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/scripts/process-css-properties.py:
* Source/WebCore/css/scripts/test/TestCSSProperties.json:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/CSSPropertyInitialValuesGeneratedInlines.h:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/CSSPropertyNames.gperf:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/CSSPropertyNames.h:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/CSSPropertyParsing.cpp:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/CSSStyleProperties+PropertyNames.idl:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/RenderStyleProperties+GettersInlines.h:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/RenderStyleProperties+SettersInlines.h:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/RenderStyleProperties.h:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/StyleBuilderGenerated.cpp:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/StyleChangedAnimatablePropertiesGenerated.cpp:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/StyleComputedStyleProperties+GettersInlines.h:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/StyleComputedStyleProperties+InitialInlines.h:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/StyleComputedStyleProperties+SettersInlines.h:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/StyleComputedStyleProperties.h:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/StyleExtractorGenerated.cpp:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/StyleInterpolationWrapperMap.cpp:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/StylePropertyShorthandFunctions.cpp:
* Source/WebCore/style/StyleBuilderCustom.h:
(WebCore::Style::BuilderCustom::applyValueDirection): Deleted.
(WebCore::Style::BuilderCustom::applyValueWritingMode): Deleted.
(WebCore::Style::BuilderCustom::applyValueColorScheme): Deleted.
(WebCore::Style::BuilderCustom::applyInitialBorderBottomLeftRadius): Deleted.
(WebCore::Style::BuilderCustom::applyInheritBorderBottomLeftRadius): Deleted.
(WebCore::Style::BuilderCustom::applyValueBorderBottomLeftRadius): Deleted.
(WebCore::Style::BuilderCustom::applyInitialBorderBottomRightRadius): Deleted.
(WebCore::Style::BuilderCustom::applyInheritBorderBottomRightRadius): Deleted.
(WebCore::Style::BuilderCustom::applyValueBorderBottomRightRadius): Deleted.
(WebCore::Style::BuilderCustom::applyInitialBorderTopLeftRadius): Deleted.
(WebCore::Style::BuilderCustom::applyInheritBorderTopLeftRadius): Deleted.
(WebCore::Style::BuilderCustom::applyValueBorderTopLeftRadius): Deleted.
(WebCore::Style::BuilderCustom::applyInitialBorderTopRightRadius): Deleted.
(WebCore::Style::BuilderCustom::applyInheritBorderTopRightRadius): Deleted.
(WebCore::Style::BuilderCustom::applyValueBorderTopRightRadius): Deleted.
(WebCore::Style::BuilderCustom::applyValueStrokeWidth): Deleted.
(WebCore::Style::BuilderCustom::applyValueStrokeColor): Deleted.
(WebCore::Style::BuilderCustom::applyInitialColor): Deleted.
(WebCore::Style::BuilderCustom::applyInheritColor): Deleted.
(WebCore::Style::BuilderCustom::applyInitialPaddingBottom): Deleted.
(WebCore::Style::BuilderCustom::applyInheritPaddingBottom): Deleted.
(WebCore::Style::BuilderCustom::applyValuePaddingBottom): Deleted.
(WebCore::Style::BuilderCustom::applyInitialPaddingLeft): Deleted.
(WebCore::Style::BuilderCustom::applyInheritPaddingLeft): Deleted.
(WebCore::Style::BuilderCustom::applyValuePaddingLeft): Deleted.
(WebCore::Style::BuilderCustom::applyInitialPaddingRight): Deleted.
(WebCore::Style::BuilderCustom::applyInheritPaddingRight): Deleted.
(WebCore::Style::BuilderCustom::applyValuePaddingRight): Deleted.
(WebCore::Style::BuilderCustom::applyInitialPaddingTop): Deleted.
(WebCore::Style::BuilderCustom::applyInheritPaddingTop): Deleted.
(WebCore::Style::BuilderCustom::applyValuePaddingTop): Deleted.
* Tools/Scripts/webkitpy/style/checkers/jsonchecker.py:

Canonical link: https://commits.webkit.org/305011@main
